sympy/solvers/solvers.py

```diff
@@ -1557,7 +1557,22 @@ def _expand(p):
                         try:
                             t = Dummy('t')
                             iv = _solve(u - t, symbol, **flags)
+                            temp_soln = soln
                             soln = list(ordered({i.subs(t, s) for i in iv for s in soln}))
+                            count_nan = 0
+                            if S.NaN in soln:
+                                count_nan = soln.count(S.NaN)
+                                # try getting a solution for the equation
+                            if count_nan != 0:
+                                # heuristic -01(check whether the root satisfies the equation)
```

Comment:
Based on the value of `count_nan` try getting solution for the equation. Different heuristics can be applied to get the solution of the equation.
Like `heuristic -01` is: if any of the root is also the solution of the original equation `_solve` will have a check and if it finds the solution it will append it to `soln` and also decrements `count_nan` by one.

If this heuristic is not able to get to the solution we can include other heuristics accordingly(in accordance with `count_nan` variable)
(for the present case this heuristic works).